presto-main/src/main/java/com/facebook/presto/operator/SpillableGroupedTopNBuilder.java

```diff
@@ -0,0 +1,303 @@
+    public void close()
+    {
+        try (Closer closer = Closer.create()) {
+            if (inMemoryGroupedTopNBuilder != null) {
+                closer.register(inMemoryGroupedTopNBuilder::close);
+            }
+
+            if (mergingInMemoryGroupedTopNBuilder != null) {
+                closer.register(mergingInMemoryGroupedTopNBuilder::close);
+            }
+            spiller.ifPresent(closer::register);
+            closer.register(() -> localUserMemoryContext.setBytes(0));
+            closer.register(() -> localRevocableMemoryContext.setBytes(0));
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public ListenableFuture<?> startMemoryRevoke()
+    {
+        checkState(spillInProgress.isDone());
+        if (producingOutput) {
+            // All revocable memory has been released in buildResult method.
+            // At this point, InMemoryGroupedTopNBuilder is no longer accepting any input so no point in spilling.
+            return NOT_BLOCKED;
+        }
+        spillToDisk();
+        return spillInProgress;
+    }
+
+    public void finishMemoryRevoke()
+    {
+        if (spiller.isPresent()) {
+            checkState(spillInProgress.isDone());
+            spiller.get().commit();
+        }
+        this.localRevocableMemoryContext.setBytes(0);
+    }
+
+    @VisibleForTesting
+    private WorkProcessor<Page> getFinalResult(List<WorkProcessor<Page>> sortedPageStreams)
+    {
+        mergeHashSort = Optional.of(new MergeHashSort(operatorContext.aggregateUserMemoryContext()));
+        WorkProcessor<Page> mergedSortedPages = mergeHashSort.get().merge(
+                partitionTypes,
+                partitionChannels,
+                sourceTypes,
+                sortedPageStreams,
+                operatorContext.getDriverContext().getYieldSignal());
+
+        // Create final result by re-processing the sorted stream page-at-a-time through a GroupedTopNBuilder
+        return mergedSortedPages.flatTransform(new WorkProcessor.Transformation<Page, WorkProcessor<Page>>()
+        {
+            boolean reset = true;
+            long memorySize;
+
+            public WorkProcessor.TransformationState<WorkProcessor<Page>> process(Optional<Page> inputPageOptional)
+            {
+                if (reset) {
+                    initializeMergingGroupedTopNBuilder();
+                    memorySize = 0;
+                    reset = false;
+                }
+
+                boolean inputIsPresent = inputPageOptional.isPresent();
+                if (!inputIsPresent && memorySize == 0) {
+                    // no more pages and aggregation builder is empty
+                    return WorkProcessor.TransformationState.finished();
+                }
+
+                if (inputIsPresent) {
+                    Page inputPage = inputPageOptional.get();
+                    boolean done = mergingInMemoryGroupedTopNBuilder.processPage(inputPage).process();
```

Comment:
What is the reason for disabling yielding for memory reservation ? The same is true for SpillableHashAggregationBuilder. Need to investigate.